### PR TITLE
embed: fix *grpc.Server panic on GracefulStop with TLS-enabled server

### PR DESCRIPTION
**Pull request by gyuho**
Replace https://github.com/coreos/etcd/pull/8986.
Fix https://github.com/coreos/etcd/issues/8916.

### CODE REVIEW

**Comment by gyuho** [comment]
@jamesdphillips @xiang90 Semaphore CI in previous PR was not working for some reason. Creating a fresh branch for proper testing.

File: embed/etcd.go [hunk 1]
```diff
@@ -82,7 +82,7 @@ type Etcd struct {
 type peerListener struct {
 	net.Listener
 	serve func() error
-	close func(context.Context) error
+	close func(time.Duration) error
```
Comment (xiang90) [inline embed/etcd.go]: https://golang.org/pkg/net/http/#Server.Shutdown takes context. i am not sure why we want to change to time.duration.

File: embed/etcd.go [hunk 2]
```diff
@@ -219,23 +219,30 @@ func (e *Etcd) Config() Config {
 	return e.cfg
 }
 
+// Close gracefully shuts down all servers/listeners.
```
Comment (xiang90) [inline embed/etcd.go]: we should mention there is a default timeout for the graceful shutdown. or we will do a force shutdown.

File: embed/etcd.go [hunk 3]
```diff
@@ -248,32 +255,46 @@ func (e *Etcd) Close() {
...
 
-func (e *Etcd) stopGRPCServer(gs *grpc.Server) {
-	timeout := 2 * time.Second
-	if e.Server != nil {
-		timeout = e.Server.Cfg.ReqTimeout()
+func stopServers(ss *servers, timeout time.Duration) {
+	shutdownNow := func() {
+		// first, close the http.Server
+		ctx, cancel := context.WithTimeout(context.Background(), timeout)
+		ss.http.Shutdown(ctx)
```
Comment (gyuho) [inline embed/etcd.go]: @xiang90 Here we construct the context for `Shutdown`?

File: embed/etcd.go [hunk 4]
```diff
@@ -248,32 +255,46 @@ func (e *Etcd) Close() {
...
+			e.Peers[i].close(time.Second)
 		}
 	}
 }
 
-func (e *Etcd) stopGRPCServer(gs *grpc.Server) {
-	timeout := 2 * time.Second
-	if e.Server != nil {
-		timeout = e.Server.Cfg.ReqTimeout()
+func stopServers(ss *servers, timeout time.Duration) {
```
Comment (xiang90) [inline embed/etcd.go]: why not take a context but time.duration?

Comment (gyuho) [reply embed/etcd.go]: Ok, that works too. Let me change to context.

File: embed/etcd.go (hunk 3 above)
Comment (xiang90) [reply embed/etcd.go]: why not pass in ctx direction? i just not sure why we want duration at the first place.

File: embed/etcd.go (hunk 2 above)
Comment (gyuho) [reply embed/etcd.go]: @xiang90 Comments are added. PTAL.

**Comment by xiang90** [comment]
lgtm if CIs all turn green.

**Comment by gyuho** [comment]
GitHub CI status polling seems outdated. All CIs passed (https://jenkins-etcd.prod.coreos.systems/job/etcd-proxy/3639/). Will merge tomorrow.

**Comment by jamesdphillips** [comment]
Late response but thank you!

**Pull request merged**